### PR TITLE
ABY3 : Bit Injection and local decomposition.

### LINKED ISSUE #154: ABY: Bit Injection (Also Required for Falcon)
## Feature Description
Implement the bit injection algorithm 

## Is your feature request related to a problem?
The Falcon protocol requires the bit injection algorithm

## Additional Context
[ABY Paper](https://eprint.iacr.org/2018/403.pdf)

### PR DESCRIPTION
**Pull request by rasswanth-s**
## Description
Implements the ABY3 bit injection ,which converts a binary sharing of single bit into an arithmetic sharing. fixes #154.

## How has this been tested?
Added Tests for the same.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/protocol/aby3/aby3.py [hunk 1]
```diff
@@ -169,3 +207,78 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
         return r_mpc, rPrime_mpc
+
+    @staticmethod
+    def local_decomposition(
+        x: ReplicatedSharedTensor, ring_size: str
+    ) -> List[ReplicatedSharedTensor]:
+        """Performs local decomposition to generate shares of shares.
+
+        Args:
+            x(ReplicatedSharedTensor) : input RSTensor.
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:218]: NIT: One empty space between variable and `(` -- this is also the case if you write a text/paper

File: src/sympc/protocol/aby3/aby3.py [hunk 2]
```diff
@@ -169,3 +207,78 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+            ring_size(str) : Ring size to generate decomposed shares in.
+
+        Returns:
+            List[ReplicatedSharedTensor]: Decomposed shares in the given ring size.
+
+        Raises:
+            ValueError: If RSTensor does not have session uuid.
+        """
+        if x.session_uuid is not None:
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:227]: MAYBE:
```
if x.session_uuid is None:
    raise ValueError(...)

session = get_session(x.session_uuid)
```

File: src/sympc/protocol/aby3/aby3.py [hunk 3]
```diff
@@ -169,3 +207,78 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+            raise ValueError("Input RSTensor should have session_uuid")
+
+        ring_size = int(ring_size)
+        tensor_type = get_type_from_ring(ring_size)
+        rank = session.rank
+        nr_parties = session.nr_parties
+
+        zero = torch.zeros(x.shares[0].shape).type(tensor_type)
+
+        shares = [[zero, zero] for i in range(nr_parties)]
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:239]: BUG?: 

Careful here. `zero` is the same reference.
Ex:
```
zero = torch.zeros((2,3))
shares = [[zero, zero]]
shares[0][0] = 20
print(shares)
shares
[[tensor([[20.,  0.,  0.], [ 0.,  0.,  0.]]), tensor([[20.,  0.,  0.], [ 0.,  0.,  0.]])]]
```
I am not sure if this is the intended behaviour (I think not).

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:239]: Good catch :+1: .Sure I will do a deep copy

File: src/sympc/protocol/aby3/aby3.py [hunk 4]
```diff
@@ -169,3 +207,77 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+        Returns:
+            arith_share (MPCTensor): Arithmetic shares of bit in input ring size.
+        """
+        args = [[share, str(ring_size)] for share in x.share_ptrs]
+
+        decompose = parallel_execution(ABY3.local_decomposition, session.parties)(args)
+
+        x1_sh, x2_sh, x3_sh = list(zip(*decompose))
+
+        x1_sh = list(map(lambda pt: pt.resolve_pointer_type(), x1_sh))
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:271]: MAYBE: Better use list comprehension here:
[pt.resolve_pointer_type() for pr in x1_sh]

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:271]: Good suggestion , list comprehension would make it simple :+1: .

**Review by kamathhrishi — COMMENTED** [review]
A test for local decomposition would be good.

File: src/sympc/protocol/aby3/aby3.py [hunk 5]
```diff
@@ -169,3 +206,78 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+        for i in range(nr_parties):
+            rst = x.clone()
+            rst.shares = shares[i]
+            rst.ring_size = ring_size
+            rst_list.append(rst)
+
+        return rst_list
+
+    @staticmethod
+    def bit_injection(x: MPCTensor, session: Session, ring_size: int) -> MPCTensor:
```
Comment (kamathhrishi) [inline src/sympc/protocol/aby3/aby3.py:253]: Add a check to see if shares are a binary and single bit.

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:253]: Yeah, I was delaying that part,  thinking whether to have the sanity check in bit injection or local decomposition, having local decomposition would reduce one round complexity(primary reason), and also local decomposition is only done for binary shares.Shall I include it in local decomposition itself?

Comment (kamathhrishi) [reply src/sympc/protocol/aby3/aby3.py:253]: I would suggest you include both. It would help in debugging. We might use local decomposition somewhere else in the future or someone might add a line of code before local decomposition making the shares non-binary and wouldn't know results are erroneous.

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:253]: Hrishi, I initially thought local decomposition is needed only binary shares, but we require it to be ubiquitous (conversion from any ring size to another) for other ops. I will include the sanity check in bit injection for ring size 2.

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:253]: Done!

**Comment by rasswanth-s** [comment]
> A test for local decomposition would be good.

Done!

**Comment by rasswanth-s** [comment]
@gmuraru  @kamathhrishi  I have made the  changes requested, you could re-review it.

**Comment by rasswanth-s** [comment]
I have added --durations=50 flag to our workflows.Maybe , before merging a PR we could check the durations with log,If the tests of the current PR are compute intensive and time consuming, we could optimize and then merge.Our CI is very time consuming. @gmuraru

File: tests/sympc/protocol/aby3/aby3_test.py [hunk 6]
```diff
@@ -83,3 +85,105 @@ def test_invalid_mpc_pointer(get_clients) -> None:
...
+def test_bit_injection_prime(get_clients, security_type) -> None:
+    parties = get_clients(3)
+    falcon = Falcon(security_type=security_type)
+    session = Session(parties=parties, protocol=falcon)
+    SessionManager.setup_mpc(session)
+    ring_size = PRIME_NUMBER
+
+    bin_sh = torch.tensor([[1, 1], [0, 0]], dtype=torch.bool)
+
+    shares = [bin_sh, bin_sh, bin_sh]  # All possible combinations
```
Comment (kamathhrishi) [inline tests/sympc/protocol/aby3/aby3_test.py:100]: Hmmm, how are these all possible combinations?

File: tests/sympc/protocol/aby3/aby3_test.py [hunk 7]
```diff
@@ -83,3 +85,105 @@ def test_invalid_mpc_pointer(get_clients) -> None:
...
+def test_bit_injection_session_ring(get_clients, security_type) -> None:
+    parties = get_clients(3)
+    falcon = Falcon(security_type=security_type)
+    session = Session(parties=parties, protocol=falcon)
+    SessionManager.setup_mpc(session)
+    ring_size = session.ring_size
+
+    bin_sh = torch.tensor([[1, 1], [0, 0]], dtype=torch.bool)
+
+    shares = [bin_sh, bin_sh, bin_sh]  # All possible combinations
```
Comment (kamathhrishi) [inline tests/sympc/protocol/aby3/aby3_test.py:124]: Same as above comment.

File: tests/sympc/protocol/aby3/aby3_test.py (hunk 6 above)
Comment (rasswanth-s) [reply tests/sympc/protocol/aby3/aby3_test.py:100]: My mistake , I forgot to remove the comment,In previous commits, I did an exhaustive tests,It really slowed down the CI, so modified the tests, to include all the edge cases in one.

Comment (rasswanth-s) [reply tests/sympc/protocol/aby3/aby3_test.py:100]: https://github.com/OpenMined/SyMPC/pull/266/commits/903e0bdbecd79930f28f0c07d2ed45a48768304e#diff-0de61ae0880dba790984703ffe467b17bd550abec01dc758b826253d5202cfc2L118

Comment (rasswanth-s) [reply tests/sympc/protocol/aby3/aby3_test.py:100]: The exhaustive changes before.

**Review by kamathhrishi — APPROVED** [review]
Looks good to me.

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -82,7 +82,7 @@ jobs:
 
       - name: Run tests
         run: |
-          pytest tests/ --deselect=tests/notebooks -n 8 -x
+          pytest tests/ --deselect=tests/notebooks -n 8 -x --durations=50
```
Comment (gmuraru) [inline .github/workflows/tests.yml:85]: Q: Why this? (sometimes the tests hangs and we should stop them?)

Comment (rasswanth-s) [reply .github/workflows/tests.yml:85]: I included this to list the top 50 slowest tests, so we could ,check the efficiency of the changes added in the PR, if it drastically slow, we could request to optimize it

Comment (rasswanth-s) [reply .github/workflows/tests.yml:85]: It willl not repeat the tests george.I thought to keep it as a benchmark to prevent from slowing the CI

File: src/sympc/protocol/aby3/aby3.py [hunk 9]
```diff
@@ -204,3 +206,86 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+            ValueError: If RSTensor does not have session uuid.
+        """
+        if x.session_uuid is None:
+            raise ValueError("Input RSTensor should have session_uuid")
+
+        session = get_session(x.session_uuid)
+        ring_size = int(ring_size)
+        tensor_type = get_type_from_ring(ring_size)
+        rank = session.rank
+        nr_parties = session.nr_parties
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:233]: Q: Maybe we can "hardcode" the number of parties to 3 for this? (since ABY3 is 3 parties)
maybe have a global `Final` variable named `NR_PARTIES = 3`?
(Maybe we can also add a check method to validate that we have the correct number of parties before calling any method?)

File: src/sympc/protocol/falcon/falcon.py [hunk 10]
```diff
@@ -318,7 +318,12 @@ def mul_malicious(
             )
             mask = parallel_execution(Falcon.falcon_mask, session.parties)(args)
 
-        eps_shares, delta_shares = zip(*mask)
+        # zip on pointers is compute intensive
+        eps_shares, delta_shares = [mask[0][0], mask[1][0], mask[2][0]], [
```
Comment (gmuraru) [inline src/sympc/protocol/falcon/falcon.py:322]: Q: Can we split those into 2 lines? (also could you make a PR for SPDZ for this?)
```
eps_shares = [mask[0][0], mask[1], [0], mask[2][0]]
delta_shares = [mask[0][1], mask[1][1]. mask[2][1]]
```

File: src/sympc/protocol/aby3/aby3.py (hunk 9 above)
Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:233]: ABY3 has this mix of protocols george where some are three parties, and some are N parties(`truncation 2 algoritm`),I would have this sanity check for protocols which support exactly three parties :+1:

File: src/sympc/protocol/falcon/falcon.py (hunk 10 above)
Comment (rasswanth-s) [reply src/sympc/protocol/falcon/falcon.py:322]: Sure George :+1:

File: tests/sympc/protocol/aby3/aby3_test.py [hunk 11]
```diff
@@ -83,3 +85,105 @@ def test_invalid_mpc_pointer(get_clients) -> None:
...
+
+    shares = [bin_sh, bin_sh, bin_sh]
+    ptr_lst = ReplicatedSharedTensor.distribute_shares(shares, session, ring_size=2)
+    x = MPCTensor(shares=ptr_lst, session=session, shape=bin_sh.shape)
+
+    xbit = ABY3.bit_injection(x, session, ring_size)
+
+    ring0 = int(xbit.share_ptrs[0].get_ring_size().get_copy())
+    result = xbit.reconstruct(decode=False)
+    exp_res = x.reconstruct(decode=False).type(torch.uint8)
```
Comment (gmuraru) [inline tests/sympc/protocol/aby3/aby3_test.py:108]: Q: Could you hard code the expected result? (using `x.reconstruct` causes us to rely on `reconstruct` to be correctly implemented)

Comment (rasswanth-s) [reply tests/sympc/protocol/aby3/aby3_test.py:108]: Sure George :+1:

Comment (gmuraru) [reply tests/sympc/protocol/aby3/aby3_test.py:108]: Also for the other tests

File: src/sympc/protocol/aby3/aby3.py [hunk 12]
```diff
@@ -204,3 +206,86 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
...
+        for sh in decompose:
+            x1_sh.append(sh[0].resolve_pointer_type())
+            x2_sh.append(sh[1].resolve_pointer_type())
+            x3_sh.append(sh[2].resolve_pointer_type())
+
+        x1 = MPCTensor(shares=x1_sh, session=session, shape=x.shape)
+        x2 = MPCTensor(shares=x2_sh, session=session, shape=x.shape)
+        x3 = MPCTensor(shares=x3_sh, session=session, shape=x.shape)
+
+        arith_share = x1 ^ x2 ^ x3
```
Comment (gmuraru) [inline src/sympc/protocol/aby3/aby3.py:289]: Q: Here you can also use "+", right? (since they are in ring size 2 it should not matter)

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:289]: No george ,here it is session ring size or prime ring size,

Comment (rasswanth-s) [reply src/sympc/protocol/aby3/aby3.py:289]: `+` will not work george

**Review by gmuraru — APPROVED** [review]
LGTM!
